core/trino-main/src/main/java/io/trino/operator/MarkDistinctHash.java

```diff
@@ -51,26 +52,38 @@ public long getEstimatedSize()
 
     public Work<Block> markDistinctRows(Page page)
     {
-        return new TransformWork<>(
-                groupByHash.getGroupIds(page),
-                ids -> {
-                    BlockBuilder blockBuilder = BOOLEAN.createBlockBuilder(null, ids.getPositionCount());
-                    for (int i = 0; i < ids.getPositionCount(); i++) {
-                        if (ids.getGroupId(i) == nextDistinctId) {
-                            BOOLEAN.writeBoolean(blockBuilder, true);
-                            nextDistinctId++;
-                        }
-                        else {
-                            BOOLEAN.writeBoolean(blockBuilder, false);
-                        }
-                    }
-                    return blockBuilder.build();
-                });
+        return new TransformWork<>(groupByHash.getGroupIds(page), this::processNextGroupIds);
     }
 
     @VisibleForTesting
     public int getCapacity()
     {
         return groupByHash.getCapacity();
     }
+
+    private Block processNextGroupIds(GroupByIdBlock ids)
+    {
+        int positions = ids.getPositionCount();
+        if (positions > 1 && nextDistinctId == ids.getGroupCount()) {
+            // no new distinct positions
+            return new RunLengthEncodedBlock(BooleanType.createBlockForSingleNonNullValue(false), positions);
+        }
+        if (positions > 1 && nextDistinctId + positions == ids.getGroupCount()) {
+            // all positions are distinct
+            nextDistinctId = ids.getGroupCount();
+            return new RunLengthEncodedBlock(BooleanType.createBlockForSingleNonNullValue(true), positions);
+        }
+        byte[] distinctMask = new byte[positions];
+        for (int position = 0; position < distinctMask.length; position++) {
+            if (ids.getGroupId(position) == nextDistinctId) {
```

Comment:
I wonder if branchless version could perform better here (Consider benchmarking it).
Something like:
```
int distinctMaskValue = ids.getGroupId(position) == nextDistinctId ? 1 : 0;
distinctMask[position] = distinctMaskValue;
nextDistinctId += distinctMaskValue;
```

Reply:
I thought about doing that here, but I chose not to for a couple reasons:
1. There are no existing benchmarks to test `MarkDistinctHash` performance (which is itself, hard to isolate from the work of the `GroupByHash`). This also answers @sopel39's question- I don't have good benchmark numbers for this change
2. I don't think that the branchless (ie: `cmov` version of this code) is actually likely to be better than the branching version in practice. This is a hard to prove fact, but I suspect that the control flow here is largely predictable in most query workloads where either most inputs are or are not distinct (ie: most of the time, > 75% of branches will go in the same direction). We could contrive a benchmark that shows whatever we want in that regard, but I'm not sure it would hold for real workloads and so I chose to leave the general code shape as-is in this PR.

Reply:
> I don't think that the branchless (ie: cmov version of this code) is actually likely to be better than the branching version in practice. 

We've seen branch reduction to have big impact actually. You might want to investigate it int follow-up work (with different distinct ratios).